src/pages/storage/UploadIsoImage.tsx

```diff
@@ -0,0 +1,155 @@
+  const [file, setFile] = useState<File | null>(null);
+  const [name, setName] = useState<string>("");
+  const [isLoading, setLoading] = useState(false);
+  const [pool, setPool] = useState("");
+  const [uploadState, setUploadState] = useState<UploadState | null>(null);
+
+  useEffect(() => {
+    notify.clear();
+  }, []);
+
+  const {
+    data: storagePools = [],
+    isLoading: isStorageLoading,
+    error: storageError,
+  } = useQuery({
+    queryKey: [queryKeys.storage],
+    queryFn: () => fetchStoragePools(project?.name ?? ""),
+  });
+
+  if (storageError) {
+    notify.failure("Loading storage pools failed", storageError);
+    onCancel();
+  }
+
+  if (isStorageLoading) {
+    return <Loader />;
+  }
+
+  if (storagePools.length > 0 && !pool) {
+    setPool(storagePools[0].name);
+  }
+
+  const importIsoFile = () => {
+    if (!file) {
+      return;
+    }
+    notify.clear();
+    setLoading(true);
+    createIsoStorageVolume(
+      pool,
+      file,
+      name,
+      project?.name ?? "",
+      setUploadState
+    )
+      .then(() => {
+        onFinish(file.name, pool);
+      })
+      .catch((e) => {
+        notify.failure("ISO import failed", e);
+      })
+      .finally(() => {
+        setLoading(false);
+        setUploadState(null);
+        void queryClient.invalidateQueries([
+          queryKeys.storage,
+          pool,
+          queryKeys.volumes,
+        ]);
+      });
+  };
+
+  const changeFile = (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (e.target.files) {
+      const file = e.target.files[0];
+      setFile(file);
+      setName(file.name);
+    }
+  };
+
+  return (
+    <>
+      {notify.notification ? <NotificationRow /> : <UploadIsoImageHint />}
+      <div className={uploadState === null ? "" : "u-hide"}>
+        <Input
+          name="iso"
+          type="file"
+          id="iso-image"
+          label="ISO image"
+          onChange={changeFile}
+        />
+        <Input
+          name="name"
+          type="text"
+          id="name"
+          label="Name"
+          value={name}
+          onChange={(e) => setName(e.target.value)}
+          disabled={file === null}
+        />
+        <Select
+          label="Storage pool"
+          id="storagePool"
+          options={storagePools.map((pool) => ({
+            label: pool.name,
+            value: pool.name,
+          }))}
+          onChange={(e) => {
+            setPool(e.target.value);
+          }}
+          value={pool}
+          disabled={file === null}
+        />
+      </div>
+      {uploadState && (
+        <>
+          <ProgressBar percentage={Math.floor(uploadState.percentage)} />
+          <p>
+            {humanFileSize(uploadState.loaded)} loaded of{" "}
+            {humanFileSize(uploadState.total ?? 0)}
+          </p>
+          {uploadState.loaded === uploadState.total && (
+            <p>
+              <i className="p-icon--spinner u-animation--spin" /> Validating ISO
+            </p>
```

Comment:
Maybe we can make the code a bit clearer here by using our `Loader` component?
```suggestion
            <Loader text="Validating ISO" />
```
If needed, you can customise the `u-loader` class to remove the margins in this context.